File: GridView/CheckBoxInHeader/CheckBoxInHeader_VB/CheckBoxHeaderCell.vb
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports Telerik.WinControls.UI
Imports Telerik.WinControls
Imports System.Drawing

Public Class CheckBoxHeaderCell
    Inherits GridHeaderCellElement
    Private checkbox As RadCheckBoxElement

    Protected Overrides ReadOnly Property ThemeEffectiveType() As Type
        Get
            Return GetType(GridHeaderCellElement)
        End Get
    End Property

    Public Sub New(ByVal column As GridViewColumn, ByVal row As GridRowElement)
        MyBase.New(column, row)

    End Sub

    Public Overrides Sub Initialize(ByVal column As GridViewColumn, ByVal row As GridRowElement)
        MyBase.Initialize(column, row)
        column.AllowSort = False
    End Sub


    Public Overloads Overrides Sub SetContent()
    End Sub

    Protected Overrides Sub DisposeManagedResources()
        RemoveHandler checkbox.ToggleStateChanged, AddressOf checkbox_ToggleStateChanged
        MyBase.DisposeManagedResources()
    End Sub

    Protected Overloads Overrides Sub CreateChildElements()
        MyBase.CreateChildElements()
        checkbox = New RadCheckBoxElement()
        AddHandler checkbox.ToggleStateChanged, AddressOf checkbox_ToggleStateChanged
        Me.Children.Add(checkbox)
    End Sub

    Protected Overloads Overrides Function ArrangeOverride(ByVal finalSize As SizeF) As SizeF
        Dim size As SizeF = MyBase.ArrangeOverride(finalSize)

        Dim rect As RectangleF = GetClientRectangle(finalSize)
        Me.checkbox.Arrange(New RectangleF((finalSize.Width - Me.checkbox.DesiredSize.Width) / 2, (rect.Height - 20) / 2, 20, 20))

        Return size
    End Function

    Public Overloads Overrides Function IsCompatible(ByVal data As Telerik.WinControls.UI.GridViewColumn, ByVal context As Object) As Boolean
        Return data.Name = "Select" AndAlso TypeOf context Is GridTableHeaderRowElement AndAlso MyBase.IsCompatible(data, context)
    End Function

    Private Sub checkbox_ToggleStateChanged(ByVal sender As Object, ByVal args As StateChangedEventArgs)
        If TypeOf Me.ColumnInfo Is GridViewCheckBoxColumn Then
            Dim valueState As Boolean = False

            If args.ToggleState = Telerik.WinControls.Enumerations.ToggleState.[On] Then
                valueState = True
            End If

            For i As Integer = 0 To Me.ViewTemplate.RowCount - 1
                Me.ViewTemplate.Rows(i).Cells(Me.ColumnIndex).Value = valueState
            Next
        End If
    End Sub
End Class

File: GridView/CheckBoxInHeader/CheckBoxInHeader_csharp/CheckBoxHeaderCell.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Telerik.WinControls.Enumerations;
using Telerik.WinControls.UI;

namespace CheckBoxInHeader_csharp
{
    public class CheckBoxHeaderCell : GridHeaderCellElement
    {
        RadCheckBoxElement checkbox;

        protected override Type ThemeEffectiveType
        {
            get
            {
                return typeof(GridHeaderCellElement);
            }
        }

        public CheckBoxHeaderCell(GridViewColumn column, GridRowElement row)
            : base(column, row)
        {

        }

        public override void Initialize(GridViewColumn column, GridRowElement row)
        {
            base.Initialize(column, row);
            column.AllowSort = false;
        }

        public override void SetContent()
        {
        }

        protected override void DisposeManagedResources()
        {
            checkbox.ToggleStateChanged -= new StateChangedEventHandler(checkbox_ToggleStateChanged);
            base.DisposeManagedResources();
        }

        protected override void CreateChildElements()
        {
            base.CreateChildElements();
            checkbox = new RadCheckBoxElement();
            checkbox.ToggleStateChanged += new StateChangedEventHandler(checkbox_ToggleStateChanged);
            this.Children.Add(checkbox);
        }

        protected override SizeF ArrangeOverride(SizeF finalSize)
        {
            SizeF size = base.ArrangeOverride(finalSize);

            RectangleF rect = GetClientRectangle(finalSize);
            this.checkbox.Arrange(new RectangleF((finalSize.Width - this.checkbox.DesiredSize.Width) / 2, (rect.Height - 20) / 2, 20, 20));

            return size;
        }

        public override bool IsCompatible(GridViewColumn data, object context)
        {
            return data.Name == "Select" && context is GridTableHeaderRowElement
                && base.IsCompatible(data, context);
        }

        private void checkbox_ToggleStateChanged(object sender, StateChangedEventArgs args)
        {
            if (!suspendProcessingToggleStateChanged)
            {
                bool valueState = false;

                if (args.ToggleState == Telerik.WinControls.Enumerations.ToggleState.On)
                {
                    valueState = true;
                }
                this.GridViewElement.EditorManager.EndEdit();
                this.TableElement.BeginUpdate();
                for (int i = 0; i < this.ViewInfo.Rows.Count; i++)
                {
                    this.ViewInfo.Rows[i].Cells[this.ColumnIndex].Value = valueState;
                }

                this.TableElement.EndUpdate(false);

                this.TableElement.Update(GridUINotifyAction.DataChanged);

            }
        }

        private bool suspendProcessingToggleStateChanged;
        public void SetCheckBoxState(Telerik.WinControls.Enumerations.ToggleState state)
        {
            suspendProcessingToggleStateChanged = true;
            this.checkbox.ToggleState = state;
            suspendProcessingToggleStateChanged = false;
        }

        public override void Attach(GridViewColumn data, object context)
        {
            base.Attach(data, context);
            this.GridControl.ValueChanged += new EventHandler(GridControl_ValueChanged);
        }

        public override void Detach()
        {
            if (this.GridControl != null)
            {
                this.GridControl.ValueChanged -= GridControl_ValueChanged;
            }

            base.Detach();
        }

        void GridControl_ValueChanged(object sender, EventArgs e)
        {
            RadCheckBoxEditor editor = sender as RadCheckBoxEditor;
            if (editor != null)
            {
                this.GridViewElement.EditorManager.EndEdit();
                if ((ToggleState)editor.Value == ToggleState.Off)
                {
                    SetCheckBoxState(ToggleState.Off);
                }
                else if ((ToggleState)editor.Value == ToggleState.On)
                {
                    bool found = false;
                    foreach (GridViewRowInfo row in this.ViewInfo.Rows)
                    {
                        if (row != this.RowInfo && row.Cells[this.ColumnIndex].Value == null || !(bool)row.Cells[this.ColumnIndex].Value)
                        {
                            found = true;
                            break;
                        }
                    }
                    if (!found)
                    {
                        SetCheckBoxState(ToggleState.On);
                    }
                }
            }
        }
    }
}
